File: src/main/scala/com/wixpress/embed/mysql/MysqldProcess.scala
package com.wixpress.embed.mysql

import java.io._
import java.util

import com.wixpress.embed.mysql.config.MysqldConfig
import com.wixpress.embed.mysql.runtime.Mysqld
import com.wixpress.embed.mysql.support.LoggingSupport
import de.flapdoodle.embed.process.config.IRuntimeConfig
import de.flapdoodle.embed.process.distribution.Distribution
import de.flapdoodle.embed.process.extract.IExtractedFileSet
import de.flapdoodle.embed.process.io.{LogWatchStreamProcessor, Processors, StreamToLineProcessor}
import de.flapdoodle.embed.process.runtime.{AbstractProcess, ProcessControl}

import scala.collection.JavaConversions._
import scala.io.BufferedSource
import scala.util.{Success, Try}

/**
 * @author viliusl
 * @since 18/09/14
 */
class MysqldProcess(
    distribution: Distribution,
    config: MysqldConfig,
    runtimeConfig: IRuntimeConfig,
    val executable: MysqldExecutable)
  extends AbstractProcess[MysqldConfig, MysqldExecutable, MysqldProcess](distribution, config, runtimeConfig, executable) with LoggingSupport {

  var stopped = false
  val postProcessor = runtimeConfig.getCommandLinePostProcessor().process(
    distribution,
    getCommandLine(distribution, config, executable.getFile()))

  val processBuilder = ProcessControl.newProcessBuilder(postProcessor, true);

  val process = ProcessControl.start(config.supportConfig(), processBuilder);

  override def onBeforeProcess(runtimeConfig: IRuntimeConfig): Unit = {
    super.onBeforeProcess(runtimeConfig)
    initDatabase()
  }

  override def getCommandLine(distribution: Distribution, config: MysqldConfig, exe: IExtractedFileSet): util.List[String] = {
    Mysqld.getCommandLine(config, exe, pidFile(exe.executable()))
  }

  override def stopInternal(): Unit = {
    this synchronized {
      if (!stopped) {
        stopped = true

        log.info("try to stop mysqld")
        stopUsingMysqldadmin().orElse {
          log.warn("could not stop mysqld via mysqladmin, try next")
          if (!sendKillToProcess) {
            log.warn("could not stop mysqld, try next")
            if (!sendTermToProcess) {
              log.warn("could not stop mysqld, try next")
              if (!tryKillToProcess) {
                log.warn("could not stop mysqld the second time, try one last thing")
              }
            }
          }
          stopProcess
          Success()
        }
      }
    }
  }

  override def cleanupInternal(): Unit = {
  }

  override def onAfterProcessStart(process: ProcessControl, runtimeConfig: IRuntimeConfig): Unit = {
    val logWatch: LogWatchStreamProcessor = new LogWatchStreamProcessor(
      "bin/mysqld: ready for connections",
      Set[String]("[ERROR]"),
      StreamToLineProcessor.wrap(runtimeConfig.getProcessOutput.getOutput))

    Processors.connect(process.getReader, logWatch)
    Processors.connect(process.getError, logWatch)

    logWatch.waitForResult(config.timeout)

    if (!logWatch.isInitWithSuccess)
      throw new RuntimeException("mysql start failed with error: " + logWatch.getFailureFound)
    else
      setProcessId(AbstractProcess.getPidFromFile(pidFile))
  }

  override def pidFile(executableFile: File): File = {
    new File(s"${executableFile.getAbsolutePath}.pid")
  }

  private def stopUsingMysqldadmin(): Try[Unit] = Try {
    val p = Runtime.getRuntime.exec(Array[String](
      "bin/mysqladmin",
      "-uroot",//user, should be different if auth method is different, password is needed as well
      "-hlocalhost",
      "--protocol=tcp",
      //"--count=3", "--sleep=3",//try 3 times with 3 seconds interval
      s"--port=${config.port}",
      "shutdown"),
      Array[String](),
      executable.extractedFiles.generatedBaseDir())

    val retCode = p.waitFor()

    new BufferedSource(p.getInputStream).getLines().foreach(l => log.trace(s"bin/mysqladmin out: $l"))
    new BufferedSource(p.getErrorStream).getLines().foreach(l => log.trace(s"bin/mysqladmin err: $l"))

    if (retCode != 0) throw new RuntimeException(s"bin/mysqladmin stop command exited with error code: $retCode")
  }

  def initDatabase(): Unit = {
    val p = Runtime.getRuntime.exec(Array[String](
      "scripts/mysql_install_db",
      "--force", // do not lookup dns - no need for resolveip command to be present
      "--no-defaults"), // do not read defaults file.
      Array[String](),
      executable.extractedFiles.generatedBaseDir())

    val retCode = p.waitFor()

    new BufferedSource(p.getInputStream).getLines().foreach(line => log.trace(s"scripts/mysql_install_db out: $line"))
    new BufferedSource(p.getErrorStream).getLines().foreach(line => log.trace(s"scripts/mysql_install_db err: $line"))

    if (retCode != 0) throw new RuntimeException(s"'scripts/mysql_install_db' command exited with error code: $retCode")
  }
}

File: src/main/java/com/wixpress/embed/mysql/MysqldProcess.java
package com.wixpress.embed.mysql;

import com.wixpress.embed.mysql.config.MysqldConfig;
import de.flapdoodle.embed.process.collections.Collections;
import de.flapdoodle.embed.process.config.IRuntimeConfig;
import de.flapdoodle.embed.process.distribution.Distribution;
import de.flapdoodle.embed.process.distribution.Platform;
import de.flapdoodle.embed.process.extract.IExtractedFileSet;
import de.flapdoodle.embed.process.io.*;
import de.flapdoodle.embed.process.runtime.AbstractProcess;
import de.flapdoodle.embed.process.runtime.ProcessControl;

import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * @author viliusl
 * @since 27/09/14
 */
public class MysqldProcess extends AbstractProcess<MysqldConfig, MysqldExecutable, MysqldProcess> {

    private final Logger log = Logger.getLogger(this.getClass().getName());
    private boolean stopped = false;
    private final int timeout;
    final MysqldExecutable executable;
    final MysqldConfig config;

    public MysqldProcess(
            final Distribution distribution,
            final MysqldConfig config,
            final IRuntimeConfig runtimeConfig,
            final MysqldExecutable executable,
            final int timeout) throws IOException {
        super(distribution, config, runtimeConfig, executable);
        this.timeout = timeout;
        this.executable = executable;
        this.config = config;
    }

    @Override
    protected List<String> getCommandLine(Distribution distribution, MysqldConfig config, IExtractedFileSet exe) throws IOException {
        final String baseDir = exe.generatedBaseDir().getAbsolutePath();

        return Collections.newArrayList(
                exe.executable().getAbsolutePath(),
                String.format("--basedir=%s", baseDir),
                String.format("--datadir=%s/data", baseDir),
                String.format("--plugin-dir=%s/lib/plugin", baseDir),
                String.format("--pid-file=%s.pid", pidFile(exe.executable())),
                String.format("--port=%s", config.getPort()),
                "--console");//windows specific
    }

    @Override
    protected void stopInternal() {
        synchronized (this) {
            if (!stopped) {
                stopped = true;

                log.info("try to stop mysqld");
                if (!stopUsingMysqldadmin()) {
                    log.warning("could not stop mysqld via mysqladmin, try next");
                    if (!sendKillToProcess()) {
                        log.warning("could not stop mysqld, try next");
                        if (!sendTermToProcess()) {
                            log.warning("could not stop mysqld, try next");
                            if (!tryKillToProcess()) {
                                log.warning("could not stop mysqld the second time, try one last thing");
                            }
                        }
                    }
                    stopProcess();
                }
            }
        }

    }

    @Override
    protected void cleanupInternal() {

    }

    @Override
    public void onAfterProcessStart(final ProcessControl process, final IRuntimeConfig runtimeConfig) throws IOException {
        Set<String> errors = new HashSet<String>();
        errors.add("[ERROR]");
        LogWatchStreamProcessor logWatch = new LogWatchStreamProcessor(
                "ready for connections",
                errors,
                StreamToLineProcessor.wrap(runtimeConfig.getProcessOutput().getOutput()));

        Processors.connect(process.getReader(), logWatch);
        Processors.connect(process.getError(), logWatch);

        logWatch.waitForResult(timeout);

        if (!logWatch.isInitWithSuccess()) {
            throw new RuntimeException("mysql start failed with error: " + logWatch.getFailureFound());
        }
    }


    private boolean stopUsingMysqldadmin() {
        try {
            Process p = Runtime.getRuntime().exec(new String[] {
                "bin/mysqladmin",
                "-uroot",//user, should be different if auth method is different, password is needed as well
                "-hlocalhost",
                "--protocol=tcp",
                String.format("--port=%s", config.getPort()),
                "shutdown"},
                null,
                executable.getFile().generatedBaseDir());

            Processors.connect(new InputStreamReader(p.getInputStream()), Processors.logTo(log, Level.FINER));
            Processors.connect(new InputStreamReader(p.getErrorStream()), Processors.logTo(log, Level.FINER));

            int retCode = p.waitFor();

            if (retCode != 0) {
                throw new RuntimeException(String.format("'bin/mysqladmin' stop command exited with error code: %s", retCode));
            } else
                return true;

        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}
